lib/internal/source_map/source_map_cache.js

```diff
@@ -23,19 +24,26 @@ const {
 const {
   setInternalPrepareStackTrace,
 } = require('internal/errors');
-const { getLazy } = require('internal/util');
-
-// Since the CJS module cache is mutable, which leads to memory leaks when
-// modules are deleted, we use a WeakMap so that the source map cache will
-// be purged automatically:
-const getCjsSourceMapCache = getLazy(() => {
-  const { IterableWeakMap } = require('internal/util/iterable_weak_map');
-  return new IterableWeakMap();
-});
 
-// The esm cache is not mutable, so we can use a Map without memory concerns:
-const esmSourceMapCache = new SafeMap();
-// The generated sources is not mutable, so we can use a Map without memory concerns:
+// The cached module instance can be removed from the global module registry
+// with approaches like mutating `require.cache`.
+// The `moduleSourceMapCache` exposes entries by `filename` and `sourceURL`.
+// In the case of mutated module registry, obsolete entries are removed from
+// the cache by the `moduleFinalizationRegistry`.
+const moduleSourceMapCache = new SafeMap();
+const moduleFinalizationRegistry = new SafeFinalizationRegistry((key) => {
+  const instanceRef = moduleSourceMapCache.get(key).moduleInstanceRef;
```

Comment:
What happens when this callback doesn't get called, which is what FinalizationRegistry callbacks need to prepare for?

I think it would be safer to have a strong sourceURL -> WeakRef<Module> map, and a Module -> source map data WeakMap. Then couple the first one with a FinalizationRegistry that deletes entries in the first map when the module goes away. That way when the finalization registry callback don't get called, we leak some source URLs at worst, which isn't too bad. But the metadata contains line end arrays which can be huge depending on the kind of modules being mapped.

Reply:
Thanks for the suggestion! Indeed, using a WeakMap should be preferred over the undeterministic finalization registry callback.